### PR TITLE
fixes metric hashing

### PR DESCRIPTION
**Pull request by justusschock**
# Before submitting

- [ ] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?
We need the hashing to be unique. PyTorch relies on that for children discoverability.

before `hash(AUROC()) == hash(AUROC())` evaluated to true when metrics had an empty non-tensor state (like an empty list) causing some attributes to be not appearing in the children.
Fixes # (issue).

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: torchmetrics/metric.py [hunk 1]
```diff
@@ -502,7 +502,12 @@ def _filter_kwargs(self, **kwargs: Any) -> Dict[str, Any]:
         return filtered_kwargs
 
     def __hash__(self) -> int:
-        hash_vals = [self.__class__.__name__]
+        # we need to add the id here, since PyTorch requires the hash of an module to be unique.
+        # they rely on that for children discovery
+        # (see https://github.com/pytorch/pytorch/blob/v1.9.0/torch/nn/modules/module.py#L1544)
+        # For metrics that include tensors it is not a problem,
+        # since their hash is unique based on the memory location but we cannot rely on that for every metric.
+        hash_vals = [self.__class__.__name__, id(self)]
```
Comment (tchaton) [inline torchmetrics/metric.py:510]:
```suggestion
        # we need to add the id here, since PyTorch requires a module hash to be unique.
        # Internally, PyTorch nn.Module rely on that for children discovery
        # (see https://github.com/pytorch/pytorch/blob/v1.9.0/torch/nn/modules/module.py#L1544)
        # For metrics that include tensors it is not a problem,
        # since their hash is unique based on the memory location but we cannot rely on that for every metric.
        hash_vals = [self.__class__.__name__, id(self)]
```

**Review by tchaton — APPROVED** [review]
LGMT !

File: tests/helpers/testers.py [hunk 2]
```diff
@@ -511,6 +511,20 @@ def compute(self):
...
+
+    def __init__(self):
+        super().__init__()
+        self.add_state("x", [], dist_reduce_fx=None)
+
+    def update(self):
+        pass
+
+    def compute(self):
+        pass
```
Comment (SkafteNicki) [inline tests/helpers/testers.py:525]: this is the exact same as the `DummyListMetric` below
```suggestion
```

File: tests/bases/test_hashing.py [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import pytest
+
+from tests.helpers.testers import DummyMetric, DummyNonTensorMetric
```
Comment (SkafteNicki) [inline tests/bases/test_hashing.py:3]:
```suggestion
from tests.helpers.testers import DummyMetric, DummyListMetric
```

File: tests/bases/test_hashing.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+pytest.mark.parametrize(
+    "metric_cls",
+    [
+        (DummyMetric,),
+        (DummyNonTensorMetric,),
```
Comment (SkafteNicki) [inline tests/bases/test_hashing.py:9]:
```suggestion
        (DummyListMetric,),
```

**Review by SkafteNicki — APPROVED** [review]

**Review by Borda — APPROVED** [review]
